packages/flutter/lib/src/cupertino/text_selection_toolbar.dart

```diff
@@ -252,68 +237,81 @@ class _CupertinoTextSelectionToolbarShape extends SingleChildRenderObjectWidget
 // on the necessary side.
 class _RenderCupertinoTextSelectionToolbarShape extends RenderShiftedBox {
   _RenderCupertinoTextSelectionToolbarShape(
-    this._anchor,
-    this._isAbove,
+    this._anchorAbove,
+    this._anchorBelow,
+    this._shadowColor,
     super.child,
   );
 
   @override
   bool get isRepaintBoundary => true;
 
-  Offset get anchor => _anchor;
-  Offset _anchor;
-  set anchor(Offset value) {
-    if (value == _anchor) {
+  Offset get anchorAbove => _anchorAbove;
+  Offset _anchorAbove;
+  set anchorAbove(Offset value) {
+    if (value == _anchorAbove) {
       return;
     }
-    _anchor = value;
+    _anchorAbove = value;
     markNeedsLayout();
   }
 
-  bool get isAbove => _isAbove;
-  bool _isAbove;
-  set isAbove(bool value) {
-    if (_isAbove == value) {
+  Offset get anchorBelow => _anchorBelow;
+  Offset _anchorBelow;
+  set anchorBelow(Offset value) {
+    if (value == _anchorBelow) {
       return;
     }
-    _isAbove = value;
+    _anchorBelow = value;
     markNeedsLayout();
   }
 
+  Color? get shadowColor => _shadowColor;
+  Color? _shadowColor;
+  set shadowColor(Color? value) {
+    if (value == _shadowColor) {
+      return;
+    }
+    _shadowColor = value;
+    markNeedsPaint();
+  }
+
+  bool get isAbove => anchorAbove.dy >= (child?.size.height ?? 0.0) - _kToolbarArrowSize.height * 2;
+
   @override
   void performLayout() {
     final RenderBox? child = this.child;
     if (child == null) {
       return;
     }
 
-    // The child is tall enough to have the arrow clipped out of it on both sides
-    // top and bottom. Since _kToolbarHeight includes the height of one arrow, the
-    // total height that the child is given is that plus one more arrow height.
-    // The extra height on the opposite side of the arrow will be clipped out. By
+    final BoxConstraints enforcedConstraint = constraints.loosen();
```

Comment:
I'd say this is something for another PR.

Reply:
It was reverted in this patch. Could you add that back if that doesn't break the new layout of the toolbar?

Reply:
Ah, ok - I removed it because the name `heightConstraint` suggested it was height only but indeed it enforced `minWidth` too. Will restore `minWidth`.